### DISCUSSION TITLE
Serving a dynamic NextJS website partially off S3 (caching parts of .next/ folder on s3)

### DISCUSSION THREAD

**Discussion by jleclanche** [vercel/next.js#21458] [Help]
I have a NextJS website with ~100k-1M pages. It runs off a docker container that currently contains the .next directory. Under load, I'm quickly running into file descriptor limitations, which I can work around by increasing ulimits but it's symptomatic of a larger issue: I'd like to not have NodeJS *serving* anything it doesn't absolutely have to serve.

I haven't found a lot of technical documentation on the .next folder but I clearly see the following three important directories:

- `.next/cache`: Seems to be constantly accessed locally by NextJS, so shouldn't be externalized.
- `.next/static`: AFAIUI, entirely generated at build time, so can safely be dumped into S3 after `yarn build`.
- `.next/server`: A lot of it seems to be generated on the fly, and it looks like it's served under `/_next/data/<buildId>/<path>/`, but I can't really tell if anything is actually read by NextJS (beyond proxying). I also can't quite tell if it's served directly, or if NextJS needs to serve it itself for any particular reason. The build ID missing from the local paths adds a bit of complexity to serving .next directly off a cache.

To that end, I'd like to:
1. Upload .next/static to an S3 bucket after build (edit: someone on Twitter mentions the static prefix config which looks like what I need. https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix)
2. Either mount .next/server (or at least parts of it?) to s3, or somehow tell the NextJS server to interact with S3 for the staticprops file generation.

So with that I need to confirm what the access patterns on those files, as requiring NextJS to do network requests for those files would not be ideal especially if they're accessed very often. I'm especially worried about a couple of files I don't recognize in .next/server: `init-server.js.js`, `on-error-server.js.js`, `pages-manifest.json` and `ssr-module-cache.js`. I'm also concerned NextJS may *need* to be serving the files itself somehow (eg. if it adds headers or needs to know about clients accessing those files for whatever reason).

I will also need to somehow tell the clients where those files are stored. I could do this with page rules or server config at the edge but I'd rather have NextJS send a different URL (on a different domain) altogether. An approach I'm considering for that is a [custom NextJS server](https://nextjs.org/docs/advanced-features/custom-server). 

Any thoughts on this so far?

**Comment by jleclanche** [comment]
Some initial notes mostly for myself:

`pages-manifest.json` looks to be generated at build time by the `pages-manifest-plugin`, and is used by `next export` but more critically by  `next-server/server/require.ts`. So that definitely needs to stay on disk. In fact, anything [here](https://github.com/vercel/next.js/blob/f06c58911515d980e25c33874c5f18ade5ac99df/packages/next/next-server/server/require.ts) is in the same boat. (I see a `font-manifest.json` mentioned, but there's [more constants here](https://github.com/vercel/next.js/blob/f06c58911515d980e25c33874c5f18ade5ac99df/packages/next/next-server/lib/constants.ts)).

I guess I'm a little confused as to why some of the manifests are in `.next` root, and some in `server/`. 

`init-server.js.js` (weird double extension) I could only find referenced [here](https://github.com/vercel/next.js/blob/72ce0991064eb047717b09418fa9ea9a83990d35/packages/next/next-server/server/next-server.ts#L255) without the extra .js, and seems to be behind a [config.experimental.plugins flag](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/config.ts#L79) which I have not enabled.

When looking through the config I also found out about `sprFlushToDisk`, which sort-of-kind-of helps my scenario here. I see discussion [here](https://github.com/vercel/next.js/discussions/12916) about it. But I'm not sure moving the caching to memory LRU really helps. I mean, if the cached entries can be served as-is, serving them off a CDN is still better (especially for the longer-lived items, eg. > 1hr lifetime).

**Comment by jleclanche** [comment]
One half of my question has been answered with the [`assetPrefix` config](https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix), although it's unclear whether this has to be set at build time, runtime, or both (eg if I'm setting it via an environment variable, does the value have to be there when I run `next build`, or can it just be set on `next start`).

**Comment by nag** [comment]
Hi @jleclanche, I ran into a similar problem and I'm wondering how you solved it. I'm especially interested in the S3 part. 🤔

**Reply by jleclanche** [reply]
I didn't fully solve it… staticprefix / assetPrefix config stuff helped IIRC. One issue was that all this stuff had to be set at *both* build time and runtime, which was pretty annoying to deal with for a variety of reasons.

I don't have access to the project in question anymore. If Vercel shows sign of wanting to fix this I could dig some stuff up and write more about it… however, not to be too cynical but the lack of answers or solutions to this kinda tells me it's not a priority and they would rather have people running their project off their cloud platform.

Bit of a shame. With some hindsight, if I wanted a highly efficient way to cache it I'd stuff what I can on s3 and put Fastly in front of the dynamic bits. It's still far less efficient, and much more work than it needs to be though.
